player-backends/player-backend-gstreamer-vaapi/gt-player-backend-gstreamer-vaapi.c

```diff
@@ -0,0 +1,295 @@
+/*
+ *  This file is part of GNOME Twitch - 'Enjoy Twitch on your GNU/Linux desktop'
+ *  Copyright © 2017 Vincent Szolnoky <vinszent@vinszent.com>
+ *
+ *  GNOME Twitch is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  GNOME Twitch is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with GNOME Twitch. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "gt-player-backend-gstreamer-vaapi.h"
+#include "gnome-twitch/gt-player-backend.h"
+#include <gst/gst.h>
+#include <gst/video/videooverlay.h>
+#ifdef GDK_WINDOWING_X11
+#include <gdk/gdkx.h>
+#endif
+
+#define TAG "GtPlayerBackendGstreamerVaapi"
+#include "gnome-twitch/gt-log.h"
+
+typedef struct
+{
+    GstElement* playbin;
+    GstElement* upload;
+    GstElement* video_sink;
+    GstElement* video_bin;
+    GstBus* bus;
+    GstPad* pad;
+    GstPad* ghost_pad;
+
+    GtkWidget* widget;
+
+    gchar* uri;
+
+    gdouble volume;
+    gboolean playing;
+    gdouble buffer_fill;
+} GtPlayerBackendGstreamerVaapiPrivate;
+
+static void gt_player_backend_iface_init(GtPlayerBackendInterface* iface);
+
+G_DEFINE_DYNAMIC_TYPE_EXTENDED(GtPlayerBackendGstreamerVaapi, gt_player_backend_gstreamer_vaapi, PEAS_TYPE_EXTENSION_BASE, 0,
+                               G_IMPLEMENT_INTERFACE_DYNAMIC(GT_TYPE_PLAYER_BACKEND, gt_player_backend_iface_init)
+                               G_ADD_PRIVATE_DYNAMIC(GtPlayerBackendGstreamerVaapi))
+
+enum
+{
+    PROP_0,
+    PROP_VOLUME,
+    PROP_PLAYING,
+    PROP_URI,
+    PROP_BUFFER_FILL,
+    NUM_PROPS
+};
+
+static GParamSpec* props[NUM_PROPS];
+
+static GstBusSyncReply
+gst_message_cb(GstBus* bus, GstMessage* msg, gpointer udata)
+{
+    GtPlayerBackendGstreamerVaapi* self = GT_PLAYER_BACKEND_GSTREAMER_VAAPI(udata);
+    GtPlayerBackendGstreamerVaapiPrivate* priv = gt_player_backend_gstreamer_vaapi_get_instance_private(self);
+
+    switch (GST_MESSAGE_TYPE(msg))
+    {
+        case GST_MESSAGE_BUFFERING:
+        {
+            gint perc;
+
+            gst_message_parse_buffering(msg, &perc);
+
+            perc < 100 ? gst_element_set_state(priv->playbin, GST_STATE_PAUSED)
+                : gst_element_set_state(priv->playbin, GST_STATE_PLAYING);
+            g_object_set(self, "buffer-fill", (gdouble) perc/100.0, NULL);
+            break;
+        }
+        case GST_MESSAGE_ELEMENT:
+        {
+#ifdef GDK_WINDOWING_X11
```

Comment:
You also have to check if the current backend is x11 too.

Reply:
@TingPing do you know how to do that ?